mathlibtools/lib.py

```diff
@@ -0,0 +1,345 @@
+        tgt_folder = DOT_MATHLIB if self.is_mathlib else self.directory/'_cache'
+        tgt_folder.mkdir(exist_ok=True)
+        archive = tgt_folder/(str(self.rev) + '.tar.bz2')
+        if archive.exists():
+            log.info('Cache for revision {} already exists'.format(self.rev))
+            return
+        pack(self.directory, filter(Path.exists, [self.directory/'src', self.directory/'test']), 
+             archive)
+
+    def get_cache(self, force: bool = False) -> None:
+        """Tries to get olean cache.
+
+        Will raise LeanDownloadError or FileNotFoundError if no archive exists.
+        """
+        if self.is_dirty and not force:
+            raise LeanDirtyRepo
+        if self.is_mathlib:
+            self.get_mathlib_olean()
+        else:
+            unpack_archive(self.directory/'_cache'/(str(self.rev)+'.tar.bz2'),
+                           self.directory)
+
+    @classmethod
+    def from_git_url(cls, url: str, target: str = '') -> 'LeanProject':
+        """Download a Lean project using git and prepare mathlib if needed."""
+        target = target or url.split('/')[-1].split('.')[0]
+        repo = Repo.clone_from(url, target)
+        proj = cls.from_path(Path(repo.working_dir))
+        proj.run(['leanpkg', 'configure'])
+        if 'mathlib' in proj.deps:
+            proj.get_mathlib_olean()
+        return proj
+    
+    @classmethod
+    def new(cls, path: Path = Path('.')) -> 'LeanProject':
+        """Create a new Lean project and prepare mathlib."""
+        if path == Path('.'):
+            subprocess.run(['leanpkg', 'init', path.absolute().name])
+        else:
+            subprocess.run(['leanpkg', 'new', str(path)])
+
+        proj = cls.from_path(path)
+        # Work around a leanpkg bug
+        if re.match(r'^3.[5-9].*', proj.lean_version):
+            proj.lean_version = 'leanprover-community/lean:' + proj.lean_version
+            proj.write_config()
+        proj.add_mathlib()
+        return proj
+
+    def run(self, args: List[str]) -> None:
+        """Run a command in the project directory.
+           
+           args is a list as in subprocess.run"""
+        subprocess.run(args, cwd=str(self.directory))
+
+    def build(self) -> None:
+        log.info('Building project '+self.name)
+        self.run(['leanpkg', 'build'])
+
+    def upgrade_mathlib(self) -> None:
+        """Upgrade mathlib in the project.
+
+        In case this project is mathlib, we assume we are already on the branch
+        we want.
+        """
+        if self.is_mathlib:
+            try:
+                rem = next(remote for remote in self.repo.remotes 
+                           if any('leanprover' in url 
+                                  for url in remote.urls))
+            except StopIteration:
+                log.info("Couldn't find a relevant git remote. "
+                         "You may try to git pull manually and then "
+                         "run `leanproject get-cache`”)
```

Comment:
```suggestion
                         "run `leanproject get-cache`")
```
The script errors for me immediately because of this